### DISCUSSION TITLE
API Filtering by device_type model

### DISCUSSION THREAD

**Discussion by travisjohnsonga** [netbox-community/netbox#13287] [General]
Hello all - 

Is there a way to filter the API based on the model name rather than the model ID?

In my environment, I have switches model 9200, 9300, 9500.  I'd like to grab all of these (with varying submodels) by filtering with a regex if possible.

**Comment by candlerb** [comment]
Which API - REST API? Which model are you querying - Device? What query are you currently using? What version of Netbox?

This works for me with Netbox v3.5.6: `/api/dcim/devices/?device_type=foo&device_type=bar`

Note that this matches on the *slug* of the device type, not the name. It returns all devices whose type is `foo` *or* `bar`, as documented [here](https://docs.netbox.dev/en/stable/reference/filtering/#filtering-objects). To filter on the id, you'd use `/api/dcim/devices/?device_type_id=1&device_type_id=2`

There are no regular expression filters in the REST API: see [filtering doc](https://docs.netbox.dev/en/stable/reference/filtering/).

**Comment by travisjohnsonga** [comment]
You are correct, I probably should have mentioned the API, it was late :-)

/api/dcim/devices/?limit=1000&device_type_model="c9500-40x" (matches slug)

When I use that - I get back all devices in netbox.  if I use device_type_id=8, I get just that platform as expected.

In the doc, there are some operators for strings, such as __ic which I could use for every that starts with c9, if that works.

**Reply by candlerb** [reply]
It's not `device_type_model`, it's `device_type` exactly as I gave in the working example I showed.

Unfortunately, the REST API accepts but ignores all unknown query parameters.  IMO it would be much better if it returned a 4xx error saying "query parameter `device_type_model` unknown"

**Reply by travisjohnsonga** [reply]
agreed

**Comment by mtinberg** [comment]
I don't think device_type_model is the param you are looking for, it's device_type.  One thing that bites me from time to time is that if you input a parameter that doesn't exist it is just silently ignored and it can take a while to notice a typo or thinko.  I don't know if it'd be more disruptive if there was an error return {"error": "Parameter 'device_type_model' doesn't exist, check /api/docs for list of available parameters"}


—

Mark Tinberg ***@***.***>

Division of Information Technology-Network Services

University of Wisconsin-Madison


________________________________

**Reply by travisjohnsonga** [reply]
I know - I was hoping (maybe a feature request?) to be able to use the actual model in the search rather than just the ID field.

**Reply by candlerb** [reply]
What do you mean by "actual model"?  You can use the model slug (`device_type`), or you can use the model ID (`device_type_id`).

**Reply by travisjohnsonga** [reply]
Here is a snippet example, for my purposes, I need to do different tasks with C9(2|3|5)00* models by name (if it was regex)

Work around is to get all devices, and loop through them to see if they fit that, but it would be nice to just get those models.
{
            "id": 237,
            "device_type": {
                "id": 16,
                "display": "C9500-40X",
                "manufacturer": {
                    "id": 1,
                    "display": "Cisco",
                    "name": "Cisco",
                    "slug": "cisco"
                },
                **"model": "C9500-40X",**
                "slug": "c9500-40x"
            },

**Reply by travisjohnsonga** [reply]
I also figured out I can take the query string from the UI filter and throw it in the API, and it works, so there is that option as well, it's just not very dynamic

?device_type_id=30&device_type_id=8&device_type_id=25&device_type_id=10&device_type_id=24&device_type_id=37&device_type_id=22&device_type_id=9&device_type_id=16&device_type_id=1&cf_in_dnac=False

**Reply by candlerb** [reply]
There are no regexp matching [filters](https://docs.netbox.dev/en/stable/reference/filtering/) in the REST API.

However, you don't need to loop through all devices. You can provide a single API request like this:

```
/api/dcim/devices/?limit=1000&device_type=c9500-40x&device_type=c9300-abc&device_type=c9200-xyz
```

(Those are made-up model slugs, but hopefully you get the idea)

If you don't know up-front exactly what device models you have that are in that format, then you could query a list of all device types and filter them, and then build the query.

Another possible approach is to use tags on the device type. Untested:

```
/api/dcim/devices/?limit=1000&device_type__tag=cpe
```

**Reply by travisjohnsonga** [reply]
Got it on the model now.

Should this also work though as it is a string field?  device_type__ic=c9

Per https://demo.netbox.dev/static/docs/rest-api/filtering/ it appears it 'should work' unless that slug isn't a string for some reason?

**Reply by candlerb** [reply]
I think you're asking: "why does `name__ic=foo` work, but `device_type__ic=c9` not work?

The short answer is: `device_type` is an integer foreign key (column `device_type_id` in table `dcim_device`), and the only filter [defined](https://docs.netbox.dev/en/stable/reference/filtering/#foreign-keys-other-fields) in the REST API for foreign key fields is `__n`. I've tested that `device_type_id__n=1` and `device_type__n=foo` **do** work as documented.  No others are provided.

If you want a deeper answer, then look in the [source](https://github.com/netbox-community/netbox/blob/v3.5.7/netbox/dcim/filtersets.py#L817-L1051):
* The [name](https://github.com/netbox-community/netbox/blob/v3.5.7/netbox/dcim/filtersets.py#L922-L924) filter is a [MultiValueCharFilter](https://github.com/netbox-community/netbox/blob/v3.5.7/netbox/utilities/filters.py#L58-L60)
* The [device_type](https://github.com/netbox-community/netbox/blob/v3.5.7/netbox/dcim/filtersets.py#L829-L834) filter is a `ModelMultipleChoiceFilter` (and is hard-coded to the Django ORM query `device_type__slug`)

The functionality of these is defined in the [django_filters](https://django-filter.readthedocs.io/en/stable/ref/filters.html) library.

The extensions like `ic` seem to be [defined](https://github.com/netbox-community/netbox/blob/v3.5.7/netbox/utilities/constants.py#L5-L16) only for char, selection and MAC address [fields](https://github.com/netbox-community/netbox/blob/v3.5.7/netbox/netbox/filtersets.py#L138-L144).

Also, the REST API filters aren't chainable like Django ORM filters.  That is, filters like `device_type__model=C9500-40X` or `device_type__manufacturer=cisco` aren't available.
